### PR TITLE
[Shadow] Add AOF rewrite support for module auxiliary data

### PR DESCRIPTION
**Pull request by sarthakaggarwal97**
Mirror for CI agent eval. Do not merge.

### CODE REVIEW

**Comment by sarthakaggarwal97** [comment]
## PR Summary

This PR adds AOF rewrite support for module auxiliary data, mirroring the existing RDB aux mechanism. It bumps `VALKEYMODULE_TYPE_METHOD_VERSION` to 6, adds three new fields (`aux_save_aof`, `aux_load

### Walkthrough

This PR adds AOF rewrite support for module auxiliary data, mirroring the existing RDB aux mechanism. It bumps `VALKEYMODULE_TYPE_METHOD_VERSION` to 6, adds three new fields (`aux_save_aof`, `aux_load_aof`, `aux_save_aof_triggers`) plus `VALKEYMODULE_AUX_BEFORE_AOF`/`AFTER_AOF` flags, and teaches `rewriteAppendOnlyFileRio` to emit serialized aux blobs as an internal `MODULE AUXLOAD` command that's replayed on load.

This fills a real gap: modules that store global, non-keyspace state (config, counters, metadata) have had no way to round-trip through AOF-only deployments, forcing users onto RDB preambles or accepting data loss on rewrite. The design reuses the module IO format and is symmetric with `rdbSaveModulesAux`, so module authors already familiar with RDB aux hooks should find it intuitive.

A few things worth a second look. `moduleAuxLoadCommand` calls `exit(1)` on any bad input (bad encver, unknown module, load failure, missing EOF marker) — appropriate when replayed from an AOF, but the subcommand is wired into the normal `MODULE` command table with `CMD_ADMIN`, so any authenticated admin can crash the server by sending malformed `MODULE AUXLOAD`. Worth gating this to the AOF-loading context or reporting errors instead of exiting when invoked from a live client; the `AUXLOAD` entry also shouldn't appear in `MODULE HELP` if it's truly internal. Forward-compatibility concern: an AOF written on 9.1 will not replay on older servers that don't recognize the subcommand — probably fine, but should be called out in release notes. Finally, test coverage is thin on the failure paths (module not loaded at restore, truncated/corrupt aux blob, multiple modules emitting aux) and there's no assertion that a rewritten AOF without `aof-use-rdb-preamble` still interoperates correctly when switched back on. Since this touches the module API surface and the AOF format, `@core-team` consensus and a `needs-doc-pr` follow-up on valkey-doc for the new type-methods version + `MODULE AUXLOAD` entry look warranted.

### Maintainer Checklist

- Core review: changed paths suggest `@core-team` review may be needed.
- Docs: changed paths look user-facing; check for a valkey-doc follow-up and `needs-doc-pr` label.
- Extra tests: this change touches riskier Valkey areas for `unstable`; consider the `run-extra-tests` label.
- Labels: consider applying `needs-doc-pr`, `run-extra-tests`.

File: src/module.c [hunk 1]
```diff
@@ -13937,6 +14020,84 @@ void VM_ScriptingEngineDebuggerProcessCommands(int *client_disconnected,
...
+    }
+
+    moduleType *mt = moduleTypeLookupModuleByName(type_name);
+    if (mt == NULL) {
+        serverLog(LL_WARNING,
+                  "The AOF file contains MODULE AUX data for module type '%s' "
+                  "which is not loaded. You need to load the module before "
+                  "loading the AOF file.",
+                  type_name);
+        exit(1);
```
Comment (sarthakaggarwal97) [inline src/module.c:14050]: `MODULE AUXLOAD` is registered as a normal admin subcommand (`CMD_ADMIN|CMD_NOSCRIPT`) — an authenticated admin can simply run `MODULE AUXLOAD bogus 0 0 garbage` and crash the server. Every error path in this function calls `exit(1)` which is only appropriate while `loadSingleAppendOnlyFile` is driving the fake AOF client, not when a live client invokes it.

The command needs to refuse execution outside AOF-loading context. For example, at the top of `moduleAuxLoadCommand`:

```suggestion
    if (c->id != CLIENT_ID_AOF) {
        addReplyError(c, "MODULE AUXLOAD is an internal command used during AOF loading");
        return;
    }
    char *type_name = objectGetVal(c->argv[2]);
```

Alternatively, convert the `exit(1)` calls into `addReplyError` when not called by the AOF fake client.

File: src/aof.c [hunk 2]
```diff
@@ -2461,6 +2461,8 @@ int rewriteAppendOnlyFileRio(rio *aof) {
 
     if (rewriteFunctions(aof) == C_ERR) goto werr;
 
+    if (aofRewriteModulesAux(aof, VALKEYMODULE_AUX_BEFORE_AOF) == C_ERR) goto werr;
```
Comment (sarthakaggarwal97) [inline src/aof.c:2464]: `aofRewriteModulesAux` only runs in the `else` branch of `rewriteAppendOnlyFile` (when the user has explicitly disabled `aof-use-rdb-preamble`). With the default `aof-use-rdb-preamble yes`, rewrite goes through `rdbSaveRio` → `rdbSaveModulesAux`, which still uses the old `aux_save`/`aux_save_triggers`. That means a module author who implements only `aux_save_aof`/`aux_load_aof` silently loses their data on a default-configured server — their callbacks never fire.

Either invoke `aofRewriteModulesAux` from the preamble path too, or document very clearly (and probably log a warning on module load) that `aux_save_aof` only fires for `aof-use-rdb-preamble no`. As written this is a footgun — please also add test coverage for the rdb-preamble case, since the existing tests all force `aof-use-rdb-preamble no`.

File: src/module.c [hunk 3]
```diff
@@ -13937,6 +14020,84 @@ void VM_ScriptingEngineDebuggerProcessCommands(int *client_disconnected,
...
+    }
+
+    sds data = objectGetVal(c->argv[5]);
+    size_t datalen = sdslen(data);
+
+    /* Create a buffer rio from the serialized data for reading. */
+    rio bufrio;
+    rioInitWithBuffer(&bufrio, sdsnewlen(data, datalen));
```
Comment (sarthakaggarwal97) [inline src/module.c:14065]: `c->argv[5]` is already a string object holding the serialized aux blob. `sdsnewlen(data, datalen)` duplicates it just so `rioInitWithBuffer` can own it. For a large aux payload this doubles peak memory during AOF load. You can init the rio directly over the argv sds (which lives until `freeClientArgv`), then skip the `sdsfree(bufrio.io.buffer.ptr)` at the end.

File: src/module.c [hunk 4]
```diff
@@ -7673,6 +7683,79 @@ ssize_t rdbSaveModulesAux(rio *rdb, int when) {
...
+                dictReleaseIterator(di);
+                return C_ERR;
+            }
+
+            /* Emit: MODULE AUXLOAD <type_name> <encver> <when> <data> */
+            if (rioWriteBulkCount(aof, '*', 6) == 0) goto werr_inner;
+            if (rioWriteBulkString(aof, "MODULE", 6) == 0) goto werr_inner;
+            if (rioWriteBulkString(aof, "AUXLOAD", 7) == 0) goto werr_inner;
+            if (rioWriteBulkString(aof, mt->name, strlen(mt->name)) == 0) goto werr_inner;
+            if (rioWriteBulkLongLong(aof, mt->id & 1023) == 0) goto werr_inner;
```
Comment (sarthakaggarwal97) [inline src/module.c:7741]: `mt->id & 1023` works but this mask is defined in `module.h` as `VALKEYMODULE_TYPE_ENCVER_MASK`. Using the macro matches the style used in `rdb.c:2910` callers and makes the intent (extract 10-bit encver) obvious.

```suggestion
            if (rioWriteBulkLongLong(aof, mt->id & VALKEYMODULE_TYPE_ENCVER_MASK) == 0) goto werr_inner;
```

File: src/commands/module-auxload.json [hunk 5]
```diff
@@ -0,0 +1,41 @@
+{
+    "AUXLOAD": {
+        "summary": "Loads module auxiliary data. This is an internal command used by AOF.",
+        "complexity": "O(N) where N is the size of the auxiliary data.",
+        "group": "server",
+        "since": "9.1.0",
+        "arity": 6,
+        "container": "MODULE",
+        "function": "moduleCommand",
+        "command_flags": [
```
Comment (sarthakaggarwal97) [inline src/commands/module-auxload.json:10]: Peer subcommands `LOAD`/`LOADEX`/`UNLOAD` carry `NO_ASYNC_LOADING|ADMIN|NOSCRIPT|PROTECTED`. If we keep AUXLOAD publicly callable at all, it probably deserves the same treatment — especially `PROTECTED`, which gates the command behind `enable-debug-command` and would mitigate the DoS concern on the moduleAuxLoadCommand exit() paths. `NO_ASYNC_LOADING` also matters: we don't want a replica applying a stray AUXLOAD during async loading.

File: tests/modules/testaof.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    } else {
+        if (after_str) {
+            ValkeyModule_SaveSigned(aof, 1);
+            ValkeyModule_SaveString(aof, after_str);
+        }
+    }
+}
+
+int testaof_aux_load(ValkeyModuleIO *aof, int encver, int when) {
+    assert(encver == 1);
```
Comment (sarthakaggarwal97) [inline tests/modules/testaof.c:75]: If anyone ever tweaks the registered encver or runs this test against an AOF produced by a newer version, the module will abort inside the fake AOF client instead of returning an error that the test could observe. Prefer `if (encver != 1) return VALKEYMODULE_ERR;` so the AOF-load error path is exercised cleanly rather than via `assert`.

File: tests/unit/moduleapi/testaof.tcl [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        start_server [list overrides [list loadmodule "$testmodule 3" "dir" $server_path appendonly yes aof-use-rdb-preamble no save ""] keep_persistence true] {
+            r set mykey myvalue
+            r bgrewriteaof
+            waitForBgrewriteaof r
+        }
+        # Restart without the module - should succeed because no MODULE AUXLOAD was written
+        start_server [list overrides [list "dir" $server_path appendonly yes aof-use-rdb-preamble no save ""]] {
+            assert_equal "myvalue" [r get mykey]
+        }
+    }
```
Comment (sarthakaggarwal97) [inline tests/unit/moduleapi/testaof.tcl:52]: All four tests force `aof-use-rdb-preamble no`, so this suite never exercises what happens on a default-configured server. Given the preamble path doesn't invoke `aux_save_aof` at all (see comment on aof.c), at least one test should either (a) run with the default and assert a module implementing only the new callbacks loses data — confirming the documented behavior — or (b) fail to illustrate the bug. Also, a test where the module is unloaded and the AOF has to be loaded without it would verify the "empty aux is skipped" claim in the commit message.

**Review by sarthakaggarwal97 — COMMENTED** [review]
Automated review found 7 issue(s). Highest severity: `high`.
